ql/src/java/org/apache/hadoop/hive/ql/optimizer/StatsOptimizer.java

```diff
@@ -945,7 +945,10 @@ private Long getRowCnt(
         if (!StatsUtils.areBasicStatsUptoDateForQueryAnswering(tbl, tbl.getParameters())) {
           return null;
         }
-        rowCnt = Long.valueOf(tbl.getProperty(StatsSetupConst.ROW_COUNT));
+        Map<String, String> basicStats = MetaStoreUtils.isNonNativeTable(tbl.getTTable()) ?
+          tbl.getStorageHandler().getBasicStatistics(tbl) : tbl.getParameters();
```

Comment:
Discuss:
Shoule we regard the stats is always accurate when statsSource is iceberg?
 If so, we need always to keep the configuration `iceberg.hive.keep.stats` true when statsSource is iceberg, so that we can optimization the count(*) when statsSource is iceberg by **StatsOptimizer**.  Same idea i wanted to do is https://github.com/apache/hive/pull/5215
https://github.com/apache/hive/blob/4f7200dc14acf71e189db05376baf4280742ff94/iceberg/iceberg-catalog/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java#L174

https://github.com/apache/hive/blob/4f7200dc14acf71e189db05376baf4280742ff94/iceberg/iceberg-catalog/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java#L225-L227

https://github.com/apache/hive/blob/4f7200dc14acf71e189db05376baf4280742ff94/ql/src/java/org/apache/hadoop/hive/ql/optimizer/StatsOptimizer.java#L945

Reply:
Oh sorry, I missed the https://github.com/apache/hive/pull/5215 
I am not very certain how accurate is the `TOTAL_RECORDS_PROP` from the snapshot summary especially when there are deletes. 
Since we have a statsSource flag I just wanted to be consistent where we take a stats.

Reply:
yes, i think `iceberg.hive.keep.stats` should be enabled when stats source is not iceberg

Reply:
@zhangbutao, could you please check the comments in https://github.com/apache/hive/pull/5215 and maybe incorporate changes from this PR into yours

Reply:
> I am not very certain how accurate is the TOTAL_RECORDS_PROP from the snapshot summary especially when there are deletes.

Iceberg table with equal delets should not be optimized by the `count `query optimization, so we can skip to get the stats or return null in case of existing deletes.

> yes, i think iceberg.hive.keep.stats should be enabled when stats source is not iceberg

`iceberg.hive.keep.stats` should be always enabled(true) when stats source is iceberg.  `iceberg.hive.keep.stats` true means that the stats is accurate. 
HMS stats for iceberg can be not accurate if some other engines(Spark、Trino) write the table but not update the HMS stats.
But if the statsSource is iceberg, the stats is retrieved from iceberg `SnapshotSummary` which is real time and accurate.

> could you please check the comments in #5215 and maybe incorporate changes from this PR into yours

#5215 I want to optimize the `count `query with no care the values of  `iceberg.hive.keep.stats`. But i am not do the limit like your change that `not use HMS stats when statsSource is Iceberg`.  I am ok to incorporate this thange as well as some other supplements to #5215, we can continue to disscuss there.